x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskRandomTests.java

```diff
@@ -0,0 +1,259 @@
+            assertThat(status.getFollowerGlobalCheckpoint(), equalTo(testRun.finalExpectedGlobalCheckpoint));
+            assertThat(status.getIndexMetadataVersion(), equalTo(testRun.finalIndexMetaDataVerion));
+        });
+
+        task.markAsCompleted();
+        assertBusy(() -> {
+            ShardFollowNodeTask.Status status = task.getStatus();
+            assertThat(status.getNumberOfConcurrentReads(), equalTo(0));
+            assertThat(status.getNumberOfConcurrentWrites(), equalTo(0));
+        });
+    }
+
+    private ShardFollowNodeTask createShardFollowTask(int concurrency, TestRun testRun) {
+        AtomicBoolean stopped = new AtomicBoolean(false);
+        ShardFollowTask params = new ShardFollowTask(null, new ShardId("follow_index", "", 0),
+            new ShardId("leader_index", "", 0), testRun.maxOperationCount, concurrency,
+            ShardFollowNodeTask.DEFAULT_MAX_BATCH_SIZE_IN_BYTES, concurrency, 10240,
+            TimeValue.timeValueMillis(10), TimeValue.timeValueMillis(10), Collections.emptyMap());
+
+        BiConsumer<TimeValue, Runnable> scheduler = (delay, task) -> {
+            try {
+                Thread.sleep(delay.millis());
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            task.run();
+        };
+        LocalCheckpointTracker tracker = new LocalCheckpointTracker(testRun.startSeqNo - 1, testRun.startSeqNo - 1);
+        return new ShardFollowNodeTask(1L, "type", ShardFollowTask.NAME, "description", null, Collections.emptyMap(), params, scheduler) {
+
+            private volatile long indexMetadataVersion = 0L;
+            private final Map<Long, Integer> fromToSlot = new HashMap<>();
+
+            @Override
+            protected void innerUpdateMapping(LongConsumer handler, Consumer<Exception> errorHandler) {
+                handler.accept(indexMetadataVersion);
+            }
+
+            @Override
+            protected void innerSendBulkShardOperationsRequest(List<Translog.Operation> operations, LongConsumer handler,
+                                                               Consumer<Exception> errorHandler) {
+                for(Translog.Operation op : operations) {
+                    tracker.markSeqNoAsCompleted(op.seqNo());
+                }
+
+                // Emulate network thread and avoid SO:
+                Thread thread = new Thread(() -> handler.accept(tracker.getCheckpoint()));
+                thread.start();
+            }
+
+            @Override
+            protected void innerSendShardChangesRequest(long from, int maxOperationCount, Consumer<ShardChangesAction.Response> handler,
+                                                        Consumer<Exception> errorHandler) {
+
+                // Emulate network thread and avoid SO:
+                Runnable task = () -> {
+                    List<TestResponse> items = testRun.responses.get(from);
+                    if (items != null) {
```

Comment:
I think we can setup the responses on fly here. Here we can vary the number of responses and type of response.

Reply:
I prefer to generate the responses beforehand. The reason is that this will help generate a sequence of responses that the shard follow task should be able to consume successfully. For example if more than 10 errors in a row would be generated then the shard follow task would fail. We could make sure that we don't do that here, but imo that defeats the purpose of generating the responses on the fly.

Reply:
I am fine with this.